node/src/components/block_validator.rs

```diff
@@ -203,26 +211,54 @@ where
                     }
                 }
             }
-            Event::DeployFound(deploy_hash) => {
+            Event::DeployFound {
+                deploy_hash,
+                deploy_size,
+            } => {
                 // We successfully found a hash. Decrease the number of outstanding requests.
                 self.in_flight.dec(&deploy_hash);
 
+                // If a deploy is received for a given block that makes that block invalid somehow,
+                // mark it for removal.
+                let mut invalid = Vec::new();
+
+                let max_block_size = self.chainspec.deploy_config.max_block_size as usize;
+
                 // Our first pass updates all validation states, crossing off the found deploy.
-                for state in self.validation_states.values_mut() {
-                    state.missing_deploys.remove(&deploy_hash);
+                for (key, state) in self.validation_states.iter_mut() {
+                    if state.missing_deploys.remove(&deploy_hash) {
+                        let valid_new_total = state
+                            .deploy_size_running_total
+                            .checked_add(deploy_size)
+                            .filter(|new_total| *new_total <= max_block_size);
+
+                        match valid_new_total {
+                            Some(new_total) => state.deploy_size_running_total = new_total,
+                            None => {
+                                // Notify everyone still waiting on it that all is lost.
+                                info!(block=?key, %deploy_hash, %max_block_size, "block size exceeded");
+                                invalid.push(key.clone());
+                            }
+                        }
+                    }
                 }
 
                 // Now we remove all states that have finished and notify the requestors.
                 self.validation_states.retain(|key, state| {
+                    if invalid.contains(key) {
+                        state.responders.drain(..).for_each(|responder| {
+                            effects.extend(responder.respond((false, key.clone())).ignore());
+                        });
+                        return false;
```

Comment:
What's with this `return` here? Does it early exit only the current iteration of `validation_state.retain`? Usually, return is exiting from the bigger context of a method call but it's not what we want to do here or we will miss other `invalid` blocks. Instead of using `return`, can we use `if { … } else { … }` for clarity? The current version really makes me uneasy.

Reply:
Just for my own sanity, I've confirmed empirically that it works correctly here :+1: